src/validation.cpp

```diff
@@ -659,11 +659,6 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, MempoolAcceptResult& result, Works
     // Bring the best block into scope
     m_view.GetBestBlock();
 
-    // we have all inputs cached now, so switch back to dummy (to protect
-    // against bugs where we pull more inputs from disk that miss being added
-    // to coins_to_uncache)
-    m_view.SetBackend(m_dummy);
```

Comment:
Let me try to elaborate a bit: when validating packages, we have each coin in the `m_viewmempool`, which at this point is the `m_view` backend. If we set the backend to dummy, we lose the ability to look up those coins. Typically, the backend allows for going to disk (which would be unnecessary after this line and be the source of a `coins_to_uncache` leak). This is needed in `CheckSequenceLocks` right afterward, and then for `CheckInputScripts` again later.

The lines above for each input:

```c
        if (!coins_cache.HaveCoinInCache(txin.prevout)) {
            coins_to_uncache.push_back(txin.prevout);
        }
```

should catch all of the coins that we need to uncache. It's a belt-and-suspenders check I would prefer not to remove, but we would need to implement a little differently. Perhaps a boolean in `CCoinsViewMemPool` for allowing fetching from disk?

Reply:
@ariard, thanks for the push to look into this. I got a pretty simple solution: we shouldn't do `m_view.SetBackend(dummy)`, we should do `m_viewmempool.SetBackend(dummy)`, since that's the backend that's actually pointing to the coins cache.

The hierarchy is `m_view -> m_viewmempool -> coinscache`.